### PR TITLE
Make lib name optional in cargo.toml

### LINKED ISSUE #1580: Difference between Manifest format and implementation: [lib] section name key required
The manifest format (http://doc.crates.io/manifest.html#configuring-a-target) says

```
All values listed are the defaults for that option unless otherwise specified.
```
and
```toml
# The name of a target is the name of the library that will be generated. This
# is defaulted to the name of the package or project.
name = "foo"
```
---
From this I assumed that the name field is optional. But when leaving it out like this:
```
[lib]
plugin = "true"
```
I get an error from cargo
```
Executing: cargo  test
failed to parse manifest at `Cargo.toml`

Caused by:
  expected a value of type `string` for the key `lib.name`
```

### PR DESCRIPTION
**Pull request by Dineshs91**
Fixes #1580. I am a bit skeptical with the use of unwrap( ) though. Suggestion's on how to replace that would be great.

r? @alexcrichton

### CODE REVIEW

File: src/cargo/util/toml.rs [hunk 1]
```diff
@@ -389,11 +389,22 @@ impl TomlManifest {
...
+                        .. lib.clone()
+                    }
+                } else if layout.lib.is_some() && lib.name.is_none() {
+                    TomlTarget {
+                        name: Some(project.name.clone()),
+                        .. lib.clone()
+                    }
                 } else {
                     lib.clone()
                 })
```
Comment (alexcrichton) [inline src/cargo/util/toml.rs]: I think this chain of if/else can be simplified with:

```rust
TomlTarget {
    name: lib.name.clone().or(Some(project.name.clone())),
    path: lib.path.clone().or(...),
    ..lib.clone()
}
```

File: src/cargo/util/toml.rs [hunk 2]
```diff
@@ -426,28 +437,46 @@ impl TomlManifest {
...
+            Some(ref benches) => {
+                for target in benches {
+                    try!(validate_bench_name(target));
+                }
+                if benches.is_empty() {
+                    inferred_bench_targets(layout)
+                } else {
+                    benches.iter().map(|t| {
+                        t.clone()
+                    }).collect()
```
Comment (alexcrichton) [inline src/cargo/util/toml.rs]: Is this iteration + collect needed? Would a `.clone()` suffice?

File: src/cargo/util/toml.rs [hunk 3]
```diff
@@ -426,28 +437,46 @@ impl TomlManifest {
         let blacklist = vec!["build", "deps", "examples", "native"];
 
         for bin in bins.iter() {
-            if blacklist.iter().find(|&x| *x == bin.name) != None {
+            if blacklist.iter().find(|&x| *x == bin.name.as_ref().unwrap()) != None {
```
Comment (alexcrichton) [inline src/cargo/util/toml.rs]: I think that this `unwrap` can trigger a panic if the `name` is omitted on a `[[bin]]` section.

File: tests/test_cargo_test.rs [hunk 4]
```diff
@@ -681,6 +681,44 @@ test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured
...
+{compiling} syntax v0.0.1 ({dir})
+{running} target[..]syntax-[..]
+
+running 1 test
+test test ... ok
+
+test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured
+
+", compiling = COMPILING, running = RUNNING, dir = p.url())));
+});
```
Comment (alexcrichton) [inline tests/test_cargo_test.rs]: Can you expand the tests here to ensure that empty names for other kinds of targets all have first-class errors?

**Comment by alexcrichton** [comment]
I'd prefer to move all the `.name.as_ref().unwrap()` into a method called `name` on `TomlTarget` to avoid all the unwraps and provide a nicer error message as well. Once we reach the stage of unwrapping it's already verified that everything has a name, so it shouldn't ever trigger.

File: src/cargo/util/toml.rs (hunk 2 above)
Comment (Dineshs91) [inline src/cargo/util/toml.rs]: You are right. Its not required. I will fix this.

**Comment by Dineshs91** [comment]
I've made the changes and added more tests.

File: src/cargo/util/toml.rs [hunk 5]
```diff
@@ -543,21 +560,52 @@ impl TomlManifest {
...
+                Ok(())
+            }
+        },
+        None => Err(human(format!("binary target bin.name is required")))
+    }
+}
+
+fn validate_example_name(target: &TomlTarget) -> CargoResult<()> {
+    match target.name {
+        Some(_) => Ok(()),
```
Comment (Dineshs91) [inline src/cargo/util/toml.rs]: Do we have to err out for empty target names, like we are doing for binary and library targets ?

Comment (alexcrichton) [inline src/cargo/util/toml.rs]: Yeah validating example, binary, and test names the same way I think is fine (you can consolidate that here)

File: src/cargo/util/toml.rs [hunk 6]
```diff
@@ -426,28 +427,44 @@ impl TomlManifest {
...
-        } else {
-            self.bench.as_ref().unwrap().iter().map(|t| {
-                t.clone()
-            }).collect()
+        let benches = match self.bench {
+            Some(ref benches) => {
+                for target in benches {
+                    try!(validate_bench_name(target));
+                }
+                if benches.is_empty() {
```
Comment (alexcrichton) [inline src/cargo/util/toml.rs]: Could this actually pick up the same logic as the tests above? E.g. if this was explicitly specified as empty we shouldn't then to go try to infer anything.

**Comment by alexcrichton** [comment]
Just a last few minor nits, thanks @Dineshs91!

**Comment by Dineshs91** [comment]
@alexcrichton pushed changes

Thanks, you were very helpful all the way through.

**Comment by alexcrichton** [comment]
@bors: r+ 06b72059d054f021b391757675fdefb289b172c4

Thanks!

**Pull request merged**